### PR TITLE
Patch support

### PR DESCRIPTION
**Pull request by skunert**
For users who make use of patches, this PR introduces automatic patch support (nice in combination with diener for example). I check the `Cargo.toml` of the project that should be build remotely and search for `[patch.xy]` entries. Corresponding projects will then be found and transferred. A new and adjusted `Cargo.toml` is then created to reflect the new paths on the remote host and transferred via rsync.

There is a new flag `--ignore-patches` in case someone does not want to transfer patches.

Error handling is quite rough but should be enough for this small tool.

Logic:
1. Read Cargo.toml of project
2. Extract list of patches
3. For each patched crate, check if there is a path given. If not, ignore (e.g. for git patches).
4. Find the workspace of the patched crate via `cargo locate-project --workspace`
5. Add workspace to the list of projects that need to be copied
6. Modify in-memory version of Cargo.toml and point crate patch to correct remote path
7. Copy folders via rsync
8. Copy adjusted Cargo.toml via rsync

### CODE REVIEW

File: src/patches.rs [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+
+    let maybe_patches =
+        extract_patched_crates_and_adjust_toml(cargo_file_content, |p| locate_workspace_folder(p));
+
+    if let Ok(Some((patched_cargo_doc, project_list))) = maybe_patches {
+        let mut tmp_cargo_file = NamedTempFile::new()
+            .map_err(|err| format!("Unable to create a temporary file: {}", err))?;
+        tmp_cargo_file
+            .write_all(patched_cargo_doc.to_string().as_bytes())
+            .map_err(|err| format!("Unable to write temporary file: {}", err))?;
```
Comment (bkchr) [inline src/patches.rs:40]: Why not use:

```
let child = Command::new("ssh").args(&["blablab", "cat - > remote_cargo_file"]).stdin(Stdio::piped()).spawn().unwrap();

child.stdin.take().unwrap().write_all("content");

child.exit()
```

Then you don't need to create temp file and rsync it.

File: src/patches.rs [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+                        .iter_mut()
+                        .filter_map(|(_, patch_table)| patch_table.as_inline_table_mut())
+                })
+                .collect()
+        });
+
+    if patched_paths.is_none() {
+        log::debug!("No patches in project.");
+        return Ok(None);
+    }
```
Comment (bkchr) [inline src/patches.rs:120]: Maybe properly unwrap the `patched_paths` with `if let Some()`?

File: src/patches.rs [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+
+    for inline_crate_table in patched_paths.unwrap() {
+        // We only act if there is a path given for a crate
+        if let Some(path) = inline_crate_table.get("path") {
+            let path = PathBuf::from(path.as_str().unwrap().clone());
```
Comment (bkchr) [inline src/patches.rs:125]: Maybe also a map_err instead of the unwrap?

File: src/patches.rs [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+    build_server: &String,
+    patched_cargo_file: NamedTempFile,
+    projects_to_copy: Vec<PatchProject>,
+) {
+    for patch_operation in projects_to_copy.iter() {
+        let local_proj_path = format!("{}/", patch_operation.local_path.to_string_lossy());
+        let remote_proj_path = format!(
+            "{}:remote-builds/{}",
+            build_server,
+            patch_operation.name.to_string_lossy()
```
Comment (bkchr) [inline src/patches.rs:198]:
```suggestion
        let local_proj_path = format!("{}/", patch_operation.local_path.display());
        let remote_proj_path = format!(
            "{}:remote-builds/{}",
            build_server,
            patch_operation.name.display()
```

File: src/patches.rs (hunk 2 above)
Comment (skunert) [reply src/patches.rs:110]: I thought about this but did not want to add another level of identation here for readability.

**Comment by skunert** [comment]
Thanks for the suggestions @bkchr , code is shorter now and I was able to remove tempfile and camino dependencies 👍

**Comment by skunert** [comment]
Ah also there is now a fix included for some `rsync` issues on macos. The `-a` option usually includes `-p` option for copying permissions. When copying from macos apfs to remote, this led to errors like: 
```
rsync: [generator] failed to set permissions on "/some/file.rs": Function not implemented (38)
```
I removed the `-p` flag when on macos.

**Review by bkchr — APPROVED** [review]

**Pull request merged**